app/Models/Chat/Channel.php

```diff
@@ -20,40 +20,142 @@
+    ];
+
     public function messages()
     {
-        $this->hasMany(Message::class, 'channel_id');
+        $messages = $this->hasMany(Message::class, 'channel_id');
+
+        if ($this->type === self::TYPES['public']) {
+            $messages = $messages->where('timestamp', '>', Carbon::now()->subHours(config('osu.chat.public_backlog_limit')));
+        }
+
+        return $messages;
+    }
+
+    public function users()
+    {
+        // This isn't a has-many-through because the relationship is cross-database.
+        return User::whereIn('user_id', UserChannel::where('channel_id', $this->channel_id)->pluck('user_id'));
+    }
+
+    public function scopePublic($query)
+    {
+        return $query->where('type', self::TYPES['public']);
+    }
+
+    public function scopePM($query)
+    {
+        return $query->where('type', self::TYPES['pm']);
     }
 
     public function getAllowedGroupsAttribute($allowed_groups)
     {
         return array_map('intval', explode(',', $allowed_groups));
     }
 
-    public function getTypeAttribute($type)
+    public function isPM()
     {
-        return strtolower($type);
+        return $this->type === self::TYPES['pm'];
+    }
+
+    public function pmTargetFor(User $user)
+    {
+        if (!$user || !$this->isPM()) {
+            return;
+        }
+
+        return $this->users()->where('user_id', '<>', $user->user_id)->first();
     }
 
-    public function receiveMessage(User $sender, $body, $isAction = false)
+    public function receiveMessage(User $sender, string $content, bool $isAction = false)
     {
+        if (mb_strlen($content, 'UTF-8') >= config('osu.chat.message_length_limit')) {
+            throw new ChatMessageTooLongException(trans('api.error.chat.too_long'));
+        }
+
+        $sentMessages = Message::where('user_id', $sender->user_id)
+            ->join('channels', 'channels.channel_id', '=', 'messages.channel_id');
+
+        if ($this->type === self::TYPES['pm']) {
+            $limit = config('osu.chat.rate_limits.private.limit');
+            $window = config('osu.chat.rate_limits.private.window');
+            $sentMessages->where('type', self::TYPES['pm']);
+        } else {
+            $limit = config('osu.chat.rate_limits.public.limit');
+            $window = config('osu.chat.rate_limits.public.window');
+            $sentMessages->where('type', '!=', self::TYPES['pm']);
+        }
+
+        $sentMessages->where('timestamp', '>=', Carbon::now()->subSecond($window));
+
+        if ($sentMessages->count() > $limit) {
+            throw new ExcessiveChatMessagesException(trans('api.error.chat.limit_exceeded'));
+        }
+
         $message = new Message();
         $message->user_id = $sender->user_id;
-        $message->content = $body;
+        $message->content = $content;
         $message->is_action = $isAction;
         $message->channel()->associate($this);
         $message->save();
+        $message = $message->fresh();
```

Comment:
can just `refresh()`; also neither is needed if it's just for updating the `message_id`